doc/man/man3/slurm_admin.3
.TH "Slurm API" "3" "July 2002" "Morris Jette" "Slurm administrative calls"
.SH "NAME"
.LP 
\fBslurm_job\fR \- Slurm administrative calls
.SH "SYNTAX"
.LP 
#include <slurm.h>
.LP 
void \fBslurm_free_key\fR (
.br
	slurm_key_t *\fIslurm_key_ptr\fP
.br
);
.LP 
slurm_key_t *\fBslurm_get_key\fR ( );
.LP 
int \fBslurm_reconfigure\fR ( );
.LP
int \fBslurm_update_job\fR (
.br 
	job_desc_msg_t *\fIupdate_job_msg_ptr\fP
.br 
);
.LP
void \fBslurm_init_part_desc_msg\fR (
.br
	update_part_msg_t *\fIupdate_part_msg_ptr\fP 
.br 
);
.LP
int \fBslurm_update_node\fR ( 
.br 
	update_node_msg_t *\fIupdate_node_msg_ptr\fP 
.br 
);
.LP
int \fBslurm_update_partition\fR ( 
.br 
	update_part_msg_t *\fIupdate_part_msg_ptr\fP 
.br 
);
.SH "ARGUMENTS"
.LP 
.TP 
\fIslurm_key_ptr\fP
Specifies the pointer to a Slurm generated key as returned by \fBslurm_get_key\fR.
.TP
\fIupdate_job_msg_ptr\fP
Specifies the pointer to a job update request specification. See slurm.h for full details on the data structure's contents. 
.TP 
\fIupdate_node_msg_ptr\fP
Specifies the pointer to a node update request specification. See slurm.h for full details on the data structure's contents. 
.TP 
\fIupdate_part_msg_ptr\fP
Specifies the pointer to a partition update request specification. See slurm.h for full details on the data structure's contents. 
.SH "DESCRIPTION"
.LP 
\fBslurm_free_key\fR Release the storage generated in response to a call of the function \fBslurm_get_key\fR.
.LP 
\fBslurm_get_key\fR Generate a key authorizing use of some Slurm partitions (depending upon partition configuration. This storage should be released by executing \fBslurm_free_key\fR. This function may only be successfully executed by user root.
.LP
\fBslurm_init_part_desc_msg\fR Initialize the contents of a partition descriptor with default values. Execute this function before executing \fBslurm_update_part\fR.
.LP 
\fBslurm_reconfigure\fR Request that the Slurm controller re-read its configuration file. The new configuration parameters take effect immediately. This function may only be successfully executed by user root.
.LP 
\fBslurm_update_job\fR Request that the configuration of a job be updated. Note that most, but not all paramters of a job may be changed by this function. Initialize the data structure using the \fBslurm_init_job_desc_msg\fR function to avoid making unanticipated changes to a job's configuration. This function may only be successfully executed by user root.
.LP 
\fBslurm_update_node\fR Request that the state of one or more nodes be updated. Note that the state of a node (e.g. DRAINING, IDLE, etc.) may be changed, but its hardware configuration may not be changed by this function. If the hardware configuration of a node changes, update the Slurm configuration file and execute the \fBslurm_reconfigure\fR function. This function may only be successfully executed by user root.
.LP 
\fBslurm_update_part\fR Request that the configuration of a partition be updated. Note that most, but not all paramters of a partition may be changed by this function. This function may only be successfully executed by user root.
.SH "RETURN VALUE"
.LP
On success, zero is returned. On error, -1 is returned, and the Slurm error code is set appropriately.
.SH "ERRORS"
.LP
\fBSLURM_PROTOCOL_VERSION_ERROR\fR Protocol version has changed, re-link your code.
.LP
\fBESLURM_INVALID_NODE_NAME\fR the requested node name(s) is/are not valid. 
.LP
\fBESLURM_INVALID_PARTITION_NAME\fR the requested partition name is not valid. 
.LP
\fBESLURM_ACCESS_DENIED\fR the requesting user lacks authorization for the requested action (e.g. trying to delete or modify another user's job). 
.SH "EXAMPLE"
.eo
.LP 
#include <stdio.h>
.br
#include <slurm.h>
.LP 
int main (int argc, char *argv[])
.br 
{
.br 
	job_desc_msg_t update_job_msg;
.br
	update_node_msg_t update_node_msg;
.br
	partition_desc_msg_t update_part_msg ;
.br 
	resource_allocation_response_msg_t* slurm_alloc_msg_ptr ;
.br
	slurm_key_t *slurm_key_ptr;
.LP 
	if (slurm_reconfigure ( )) {
.br
		printf ("reconfigure errno %d\n", slurm_get_errno());
.br 
		exit (1);
	}
.LP 
	slurm_init_job_desc_msg( &update_job_msg );
.br 
	update_job_msg. job_id = 1234;
.br 
	update_job_msg. time_limit = 200;
.br 
	if (slurm_update_job (&update_job_msg)) { 
.br
		printf ("job update errno %d\n", slurm_get_errno());
.br 
		exit (1);
	}
.LP 
	slurm_init_part_desc_msg ( &update_part_msg );
.br
	update_part_msg . name = "test.partition";
.br
	update_part_msg . state_up = 0;	/* make partition down */
.br 
	if (slurm_update_partition (&update_part_msg)) { 
.br
		printf ("partition update errno %d\n", slurm_get_errno());
.br 
		exit (1);
	}
.LP 
	update_node_msg . node_names = "lx[10-12]";
.br
	update_node_msg . node_state = NODE_STATE_DRAINING ;
.br 
	if (slurm_update_node (&update_node_msg)) { 
.br
		printf ("node update errno %d\n", slurm_get_errno());
.br 
		exit (1);
.br 
	}
.LP
	slurm_key_ptr = slurm_get_key ( );
.br
	slurm_init_part_desc_msg ( &update_part_msg );
.br 
	job_mesg. name = ("job01\0");
.br 
	job_mesg. partition = ("reserved\0");;
.br 
	job_mesg. num_nodes = 400;
.br
	job_mesg. slurm_key_ptr = slurm_key_ptr;
.br 
	if (slurm_allocate_resources(&job_desc_msg,&slurm_alloc_msg_ptr,true)) {
.br
		printf ("allocate errno %d\n", slurm_get_errno());
.br 
		exit (1);
.br 
	}
.br 
	printf ("Allocated nodes %s to job_id %u\n", 
.br 
		slurm_alloc_msg_ptr->node_list, slurm_alloc_msg_ptr->job_id );
.br
	slurm_free_resource_allocation_response_msg ( slurm_alloc_msg_ptr );
.br 
	slurm_free_key ( slurm_key_ptr );
.br 
	exit (0);
.br 
}
.ec
.SH "COPYING"
Copyright (C) 2002 The Regents of the University of California.
Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
UCRL-CODE-2002-040.
.LP
This file is part of SLURM, a resource management program.
For details, see <http://www.llnl.gov/linux/slurm/>.
.LP
SLURM is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your option)
any later version.
.LP
SLURM is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
details.
.SH "SEE ALSO"
.LP 
\fBscontrol\fR(1), \fBslurm_allocate_resources\fR(3), \fBslurm_get_errno\fR(3), \fBslurm_init_job_desc_msg\fR(3), \fBslurm_job_will_run\fR(3), \fBslurm_perror\fR(3), \fBslurm_strerror\fR(3), \fBslurm_submit_batch_job\fR(3)


NEWS
This file describes changes in recent versions of Slurm. It primarily
documents those changes that are of interest to users and administrators.

* Changes in Slurm 21.08.0pre1
==============================
 -- slurmrestd - add v0.0.37 OpenAPI plugin.
 -- slurmrestd/v0.0.37 - rename standard_in -> standard_input.
 -- slurmrestd/v0.0.37 - rename standard_out -> standard_output.
 -- slurmdbd - Improve log messages of more time than possible on rollups.
 -- Changed the --format handling for negative field widths (left justified)
    to apply to the column headers as well as the printed fields.
 -- Add LimitFactor to the QOS. A float that is factored into an associations
    [Grp|Max]TRES limits.  For example, if the LimitFactor is 2, then an
    association with a GrpTRES of 30 CPUs, would be allowed to allocate 60
    CPUs when running under this QOS.
 -- slurmrestd - Pass SLURM_NO_CHANGE_IN_DATA to client as 403 (Not Modified).
 -- slurmrestd/v0.0.37 - Add update_time field to Jobs query to allow clients
    to only get jobs list based on change timestamp.
 -- Reset job eligible time when job is manually held.
 -- Add DEBUG_FLAG_JAG to improve logging related to job account gathering.
 -- Convert logging in account_gather/common to DEBUG_FLAG_JAG.
 -- Add more logging for jag_common_poll_data() when prec_extra() called.
 -- slurmrestd/v0.0.37 - add API to fetch reservation(s) info.
 -- Catch more errors in task/cgroup initalization and cleanup to avoid allowing
    jobs to start when cgroups failure to configure correctly.
 -- Fix cgroup ns detection when using containers (e.g. LXC or Docker).
 -- Reset job's next_step_id counter to 0 after being requeued.
 -- Make scontrol exit with non-zero status after failing to delete a partition
    or reservation.
 -- Make NtasksPerTRES optional in slurm_sprint_job_info().
 -- slurmrestd/v0.0.37 - Add update_time field to nodes query to allow clients
    to only get nodes list based on change timestamp.
 -- common/parse_config - catch and propagate return codes when handling a match
    on a key-value pattern. This implies error codes detected in the handlers
    are now not ignored and users of _handle_keyvalue_match() can fatal().
 -- common/hostlist - fix hostlist_delete_nth() xassert() upper bound check.
 -- API change: Removed slurm_kill_job_msg and modified the function signature
    for slurm_kill_job2. slurm_kill_job2 should be used instead of
    slurm_kill_job_msg.
 -- Fix non-zero exit code for scontrol ping when all controllers are down.
 -- Enforce a valid configuration for AccountingStorageEnforce in slurm.conf.
    If the configuration is invalid, then an error message will be printed and
    the command or daemon (including slurmctld) will not run.
 -- slurmrestd/v0.0.37 - Add update_time field to partitions/reservations query
    to allow clients to only get the entities list when something changed.
 -- slurmdbd.service - add "After" relationship to all common names for MariaDB
    to reduce startup delays.
 -- slurmrestd/v0.0.37 - Correct displaying node states that are UNKNOWN.
 -- slurmrestd/v0.0.37 - Add flags to node states.
 -- Fix first job on fresh cluster not being assigned JobId=1 (or FirstJobId).
 -- squeue - make it so --nodelist is sensitive to --clusters.
 -- squeue - do --nodelist node validation in the same order as listing.
 -- Removed AccountingStoreJobComment option.  Please update your config to use
    AccountingStoreFlags=job_comment instead.
 -- AccountingStoreFlags=job_script allows you to store the job's batch script.
 -- AccountingStoreFlags=job_env allows you to store the job's env vars.
 -- Add sacct -o SubmitLine to get the submit line of a job/step.
 -- Removed DefaultStorage{Host,Loc,Pass,Port,Type,User} options.
 -- Fix NtasksPerTRES delimiter from : to = in scontrol show job output.
 -- Removed CacheGroups, CheckpointType, JobCheckpointDir, MemLimitEnforce,
    SchedulerPort, SchedulerRootFilter options.
 -- Make job accounting queries use consistent timeframes with and w/o jobs.
 -- --cpus-per-task and --threads-per-core now imply --exact.
    This fixes issues where steps would be allocated the wrong number of CPUs.
 -- configure: the --with option handling has been made consistent across the
    various optional libraries. Specifying --with-foo=/path/to/foo will only
    check that directory for the applicable library (rather than, in some cases,
    falling back to the default directories), and will always error the build
    if the library is not found (instead of a mix of error messages and non-
    fatal warning messages).
 -- configure: replace --with-rmsi_dir option with proper handling for
    --with-rsmi=dir.
 -- Pass additional job environment variables to MailProg.
 -- Add SLURM_JOB_WORK_DIR to Prolog, Epilog.
 -- Removed sched/hold plugin.
 -- Fix srun overwriting SLURM_SUBMIT_DIR and SLURM_SUBMIT_HOST when within an
    existing allocation.
 -- step_ctx code has been removed from the api.
 -- cli_filter/lua, jobcomp/lua, job_submit/lua now load their scripts from the
    same directory as the slurm.conf file (and thus now will respect changes
    to the SLURM_CONF environment variable).
 -- SPANK - call slurm_spank_init if defined without slurm_spank_slurmd_exit in
    slurmd context.
 -- job_container/tmpfs - Remove need for .active file to allow salloc without
    an interactive step to work.
 -- Add new 'PLANNED' state to a node to represent when the backfill scheduler
    has it planned to be used in the future instead of showing as 'IDLE'.
 -- slurmd - Delay background node registration on every failure up to 128s on
    startup.
 -- slurmctld - Always notify slurmd that node registration was accepted to
    avoid slurmd needless attempting to re-register if there is configuration
    issue.
 -- Put node into "INVAL" state upon registering with an invalid node
    configuration. Node must register with a valid configuration to continue.

src/scontrol/Slurm_Admin.c
/*
 * Slurm_Admin - Administration tool for SLURM. 
 * Provides interface to read, write, update, and configurations.
 */

#include <errno.h>
#include <stdio.h>
#include <string.h>

#define BUF_SIZE 1024
#define DEBUG_MODULE 1
#define MAX_INPUT_FIELDS 50

char *Command_Name;
int Exit_Flag;		/* Program to terminate if =1 */
int Quiet_Flag;		/* Quiet=1, verbose=-1, normal=0 */

void Dump_Command(int argc, char *argv[]);
int Get_Command(int *argc, char *argv[]);
int Process_Command(int argc, char *argv[]);
void Usage();

main(int argc, char *argv[]) {
    int i, j;
    int Error_Code;
    int Input_Field_Count;
    char *Input_Fields[MAX_INPUT_FIELDS];

    Command_Name = argv[0];
    Exit_Flag = 0;
    Input_Field_Count = 0;
    Quiet_Flag = 0;

    if ((argc-1) > MAX_INPUT_FIELDS) {
	fprintf(stderr, "%s error: No more than %d input fields permitted\n", Command_Name, MAX_INPUT_FIELDS);
	exit(E2BIG);
    } /* if */

    if (argc > 1) {
	if (strcmp(argv[1], "-q") == 0) {
	    Quiet_Flag =  1;
	    i = 2;
	} else if (strcmp(argv[1], "-v") == 0) {
	    Quiet_Flag = -1;
	    i = 2;
	} else {
	    i = 1;
	} /* else */

	Input_Field_Count = argc - i;
	if (Input_Field_Count > 0) Exit_Flag=1;
	for (j=i; j<argc; j++) {
	    Input_Fields[j-i] = argv[j];
	} /* for */
    } /* if */
    if (Input_Field_Count == 0) Error_Code=Get_Command(&Input_Field_Count, Input_Fields);

    while (1) {
Dump_Command(Input_Field_Count, Input_Fields);
	Error_Code = Process_Command(Input_Field_Count, Input_Fields);
	if (Error_Code != 0) break;
	if (Exit_Flag == 1) break;
	Error_Code = Get_Command(&Input_Field_Count, Input_Fields);
	if (Error_Code != 0) break;
    } /* while */

    exit(Error_Code);
} /* main */


/*
 * Dump_Command - Dump the user's command
 * Input: argc - count of arguments
 *        argv - the arguments
 */
void Dump_Command(int argc, char *argv[]) {
    int i;

    for (i=0; i<argc; i++) {
	printf("Arg %d:%s:\n", i, argv[i]);
    } /* for */
} /* Dump_Command */


/*
 * Get_Command - Get a command from the user
 * Input: argc - location to store count of arguments
 *        argv - location to store the argument list
 * Output: returns error code, 0 if no problems
 */
int Get_Command(int *argc, char *argv[]) {
    static char *In_Line;
    static int In_Line_Size = 0;
    int In_Line_Pos = 0;
    int Temp_Char, i;

    if (In_Line_Size == 0) {
	In_Line_Size += BUF_SIZE;
	In_Line = (char *)malloc(In_Line_Size);
	if (In_Line == NULL) {
	    fprintf(stderr, "%s: Error %d allocating memory\n", Command_Name, errno);
	    In_Line_Size = 0;
	    return errno;
	} /* if */
    } /* if */
	
    printf("slurm_admin: ");
    *argc = 0;
    In_Line_Pos = 0;

    while (1) {
	Temp_Char = getc(stdin);
	if (Temp_Char == EOF) break;
	if (Temp_Char == (int)'\n') break;
	if ((In_Line_Pos+2) >= In_Line_Size) {
	    In_Line_Size += BUF_SIZE;
	    In_Line = (char *)realloc(In_Line, In_Line_Size);
	    if (In_Line == NULL) {
		fprintf(stderr, "%s: Error %d allocating memory\n", Command_Name, errno);
		In_Line_Size = 0;
		return errno;
	    } /* if */
	} /* if */
	In_Line[In_Line_Pos++] = (char)Temp_Char;
    } /* while */
    In_Line[In_Line_Pos] = (char)NULL;

    for (i=0; i<In_Line_Pos; i++) {
	if (isspace((int)In_Line[i])) continue;
	if (((*argc)+1) > MAX_INPUT_FIELDS) {
	    fprintf(stderr, "%s: Over %d fields in line: %s\n", Command_Name, MAX_INPUT_FIELDS ,In_Line);
	    return E2BIG;
	} /* if */
	argv[(*argc)++] = &In_Line[i];
	for (i++ ; i<In_Line_Pos; i++) {
	    if (!isspace((int)In_Line[i])) continue;
	    In_Line[i] = (char)NULL;
	    break;
	} /* for */
    } /* for */
    return 0;
} /* Get_Command */


/*
 * Process_Command - Process the user's command
 * Input: argc - count of arguments
 *        argv - the arguments
 * Ourput: Return code is 0 or errno (ONLY for errors fatal to Slurm_Admin)
 */
int Process_Command(int argc, char *argv[]) {

    if ((strcmp(argv[0], "exit") == 0) || 
        (strcmp(argv[0], "quit") == 0)) {
	if (argc > 1) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	} /* if */
	Exit_Flag = 1;
	return 0;
    } /* if exit */

   if (strcmp(argv[0], "help") == 0) {
	if (argc > 1) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	} /* if */
	Usage();
	return 0;
    } /* if */

    if (strcmp(argv[0], "quiet") == 0) {
	if (argc > 1) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	} /* if */
	Quiet_Flag = 1;
	return 0;
    } /* if */

/*  if (strcmp(argv[0], "quit") == 0)  See "exit" above */

    if (strcmp(argv[0], "reconfigure") == 0) {
	if (argc > 2) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	    return 0;
	} /* if */
	printf("%s keyword not yet implemented\n", argv[0]);
	return 0;
    } /* if */

    if (strcmp(argv[0], "restart") == 0) {
	if (argc > 2) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	    return 0;
	} /* if */
	printf("%s keyword not yet implemented\n", argv[0]);
	return 0;
    } /* if */

    if (strcmp(argv[0], "show") == 0) {
	if (argc > 3) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	    return 0;
	} /* if */
	if (argc < 2) {
	    fprintf(stderr, "Too few arguments for %s keyword\n", argv[0]);
	    return 0;
	} /* if */ 
	if ((strcmp(argv[1],"job") != 0)  && 
	    (strcmp(argv[1],"node") != 0) && (strcmp(argv[1],"partition") != 0)) {
	    fprintf(stderr, "Invalid entity %s for %s keyword\n", argv[1], argv[0]);
	    return 0;
	} /* if */
	printf("%s keyword not yet implemented\n", argv[0]);
	return 0;
    } /* if */

    if (strcmp(argv[0], "start") == 0) {
	if (argc > 2) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	    return 0;
	} /* if */
	printf("%s keyword not yet implemented\n", argv[0]);
	return 0;
    } /* if */

    if (strcmp(argv[0], "stop") == 0) {
	if (argc > 2) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	    return 0;
	} /* if */
	printf("%s keyword not yet implemented\n", argv[0]);
	return 0;
    } /* if */

    if (strcmp(argv[0], "update") == 0) {
	if (argc < 3) {
	    fprintf(stderr, "Too few arguments for %s keyword\n", argv[0]);
	    return 0;
	} /* if */
	if ((strcmp(argv[1],"job") != 0)  && 
	    (strcmp(argv[1],"node") != 0) && (strcmp(argv[1],"partition") != 0)) {
	    fprintf(stderr, "Invalid entity %s for %s keyword\n", argv[1], argv[0]);
	    return 0;
	} /* if */
	printf("%s keyword not yet implemented\n", argv[0]);
	return 0;
    } /* if */

    if (strcmp(argv[0], "upload") == 0) {
	if (argc > 2) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	    return 0;
	} /* if */
	printf("%s keyword not yet implemented\n", argv[0]);
	return 0;
    } /* if */

    if (strcmp(argv[0], "verbose") == 0) {
	if (argc > 1) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	} /* if */
	Quiet_Flag = -1;
	return 0;
    } /* if */

    if (strcmp(argv[0], "version") == 0) {
	if (argc > 1) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	} /* if */
	printf("%s version 0.1\n", Command_Name);
	return 0;
    } /* if */

    if (strcmp(argv[0], "write") == 0) {
	if (argc > 3) {
	    fprintf(stderr, "Too many arguments for %s keyword\n", argv[0]);
	    return 0;
	} /* if */
	if (argc < 3) {
	    fprintf(stderr, "Too few arguments for %s keyword\n", argv[0]);
	    return 0;
	} /* if */
	if ((strcmp(argv[1],"job") != 0)  && 
	    (strcmp(argv[1],"node") != 0) && (strcmp(argv[1],"partition") != 0)) {
	    fprintf(stderr, "Invalid entity %s for %s keyword\n", argv[1], argv[0]);
	    return 0;
	} /* if */
	printf("%s keyword not yet implemented\n", argv[0]);
	return 0;
    } /* if */

    fprintf(stderr, "Invalid keyword: %s\n", argv[0]);
    return 0;

} /* Process_Command */


/* Usage - Show the valid slurm_admin commands */
void Usage() {
    printf("%s [-q | -v] [<keyword>]\n", Command_Name);
    printf("    -q is equivalent to the keyword \"quiet\" described below.\n");
    printf("    -v is equivalent to the keyword \"verbose\" described below.\n");
    printf("    <keyword> may be omitted from the execute line and %s will execute in interactive\n");
    printf("     mode to process multiple keywords (i.e. commands). Valid <keyword> values are:\n\n");
    printf("     exit                         Terminate this command.\n");
    printf("     help                         Print this description of use.\n");
    printf("     quiet                        Print no messages other than error messages.\n");
    printf("     quit                         Terminate this command.\n");
    printf("     reconfigure [<NodeName>]     Re-read configuration files, default is all nodes.\n");
    printf("     restart [<NodeName>]         Stop and restart daemons, default is all nodes\n");
    printf("     show <entity> [<ID>]         Display state of identified entity, default is all records.\n");
    printf("     start [<NodeName>]           Start daemons as needed, default is all nodes\n");
    printf("     stop [<NodeName>]            Stop daemons, default is all nodes\n");
    printf("     update <entity> <options>    Update state of identified entity.\n");
    printf("     upload [<NodeName>]          Upload node configuration, default is from all nodes.\n");
    printf("     verbose                      Enable detailed logging.\n");
    printf("     version                      Display tool version number.\n");
    printf("     write <entity> <filename>    Write entity configuration to specified file.\n");
} /* Usage */
